tests/cxx/badinc.cc

```diff
@@ -704,6 +704,7 @@ I1_TemplateClass<I1_Class> i1_templateclass_object;
 // IWYU: I1_TemplateClass is...*badinc-i1.h
 // IWYU: I1_Class needs a declaration
 // IWYU: I1_Class is...*badinc-i1.h
+// IWYU: I1_Union is...*badinc-i1.h
```

Comment:
That's what I've asked about above. Currently, the "author intent" logic works only for those function parameters which have a non-explicit one-arg constructor _except copy- or move-constructors_. I thought that copy-ctors may be just forgotten, but it seems like it is by design, because this logic exists even [in the very first commit](https://github.com/include-what-you-use/include-what-you-use/commit/dee92b5e0acf7188d4e5bb3db6dec833a6d1ff2b#diff-4d5366b53d869b87c2b6cad93f1b0acc2fc8c5d20e28a6fbd058d5a2642b970fR712). Moreover, [you have extended it](https://github.com/include-what-you-use/include-what-you-use/commit/7b58488e95dcc25918ac7f2373ccc8b9c9d39f19) for C++11 move-constructors (issue #89).

> Or is this something that happened to work by accident and now needs explicit work to get the reexports right?

I'd say that it didn't work at all. If you have such files:
```cpp
// A.h

#pragma once

class A {};

// Fn.h

#include "A.h"

void Fn(A);

// UserFn.cpp

#include "Fn.h"
#include "A.h"

void UserFn(const A& a) {
    Fn(a);
}
```
IWYU version from the master branch suggests replacing `#include "A.h"` with fwd-declaration both in `Fn.h` and in `UserFn.cpp`. `A` is not considered for reexport. The code doesn't compile after that, of course.

Another case is C++11 braced initialization syntax, which may be used for implicit constructors without mentioning class name, like in such a function call: `Fn({1, 'a', 5.0});`. Shouldn't all implicit ctors be considered for "autocast"? Should IWYU require their full type info in function-declaring headers by default?

My personal opinion is: no. IWYU should maximize fwd-declarations in headers and `#include`s in `.cpp`-files, because it makes compilation faster and supports refactorings. And when a user writes `Fn({arguments, of, the, constructor})`, he probably knows what he want to construct and can add the corresponding `#include`.

Reply:
Btw, this PR doesn't handle list-initialization of [aggregates](https://en.cppreference.com/w/cpp/language/aggregate_initialization), only braced syntax of constructor calls.

Reply:
I agree with your sentiment; it does seem strange that implicit construction is different from implicit conversion. I don't have a clear idea what the larger-scale effects would be if we change this policy, but it does seem worth an experiment.

Anyway, that seems unrelated to the changes in this PR, and I think it looks good as a whole.

But I wonder if we could preserve the (broken) policy for copy-initialization in this PR? It's not entirely clear to me what changed it, and it seems orthogonal to the core of the PR.

Reply:
> But I wonder if we could preserve the (broken) policy for copy-initialization in this PR?

I'm biased that it is rather fixed than broken here. It is in all senses better to have an `#include` at the caller site than have no one at all.

Reply:
Fair enough, that makes sense to me too.

I somehow read the PR as being all about getting provision right, but now I see what you mean, thanks.

> A doubtful case is when the function is inline, or defined inside a class definition (which is essentially the same case).
> In that cases, the passed-by-value argument type should be #included in the function header. However, again, the
> requirement to #include the argument type header at the caller site supports refactoring when someone makes the
> function non-inline and places argument type fwd-declaration in the function header.

Yes, and same for return type. For _this_ PR I think it's fine to require them on both sides, but the ideal behavior should be to reexport them, right? It's a tradeoff between optimizing for stability or for build speed, I guess.

Reply:
> the ideal behavior should be to reexport them, right?

I don't think so now. Do you really want an `#include` for every passing-by-value function parameter in the function-defining header (because nearly all classes have copy- and/or move-ctor) instead of the forward-declarations? I think it is contrary to the IWYU goals, and the current "autocast" policy is sufficient.

It should be noted that IWYU cannot have policy to suggest fwd-declaring something on the author/header side and simultaneously to analyze the "provision" status of that thing on the user's side, because it would lead to uncompilable code, as in the example above.

Reply:
Just a few more nits, then let's get this merged.

Reply:
Oops, our messages crossed.

I think what I'm aiming for is that _if_ the function-declaring header includes the full type, the user should not have to. But my thinking on this is obviously not clear.

Reply:
If so, you have a problem like there is on the master branch: IWYU suggests to remove `#include "A.h"` both from the header `Fn.h` and from the `UserFn.h` (according to the policy in that case, because `#include` is still present in `Fn.h` at the moment of analysis), and the code is not compilable after that.